routing/DELEGATED_CONTENT_ROUTING_HTTP.md

```diff
@@ -0,0 +1,261 @@
+
+## Error Codes
+
+- `501`: must be returned if a method/path is not supported
+- `429`: may be returned to indicate to the caller that it is issuing requests too quickly
+- `400`: must be returned if an unknown path is requested
+
+## CORS and Web Browsers
+
+Browser interoperability requires implementations to support
+[CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS).
+
+JavaScript client running on a third-party Origin must be able to send HTTP
+request to the endpoints defined in this specification, and read the received
+values. This means HTTP server implementing this API must (1) support
+[CORS preflight requests](https://developer.mozilla.org/en-US/docs/Glossary/Preflight_request)
+sent as HTTP OPTIONS, and (2) always respond with headers that remove CORS
+limits, allowing every website to query the API for results:
+
+```plaintext
+Access-Control-Allow-Origin: *
+Access-Control-Allow-Methods: GET, PUT, POST, DELETE, OPTIONS
+```
+
+## Known Transfer Protocols
+This section contains a non-exhaustive list of known transfer protocols (by name) that may be supported by clients and servers.
+
+### Bitswap
+Multicodec name: `transport-bitswap`
+Schema: `bitswap`
+
+#### Read Provider Records
+
+```json
+{
+    "Protocol": "transport-bitswap",
+    "Schema": "bitswap",
+    "ID": "12D3K...",
+    "Addrs": ["/ip4/..."]
+}
+```
+
+- `ID`: the [Peer ID](https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md) to contact
+- `Addrs`: a list of known multiaddrs for the peer
+  - This list may be incomplete or incorrect and should only be treated as *hints* to improve performance by avoiding extra peer lookups
+
+The server should respect a passed `transport` query parameter by filtering against the `Addrs` list.
+
+
+#### Write Provider Records
+
+```json
+{
+    "Protocol": "transport-bitswap",
+    "Schema": "bitswap",
+    "Signature": "<signature>",
+    "Payload": "<payload>"
+}
+```
+
+- `Signature`: a multibase-encoded signature of the sha256 hash of the `Payload` field, signed using the private key of the Peer ID specified in the `Payload` JSON
+  - Servers may ignore this field if they do not require signature verification
+- `Payload`: a string containing a serialized JSON object which conforms with the following schema:
+```json
+{
+    "Keys": ["cid1", "cid2"],
+    "Timestamp": 0,
+    "AdvisoryTTL": 0,
+    "ID": "12D3K...",
+    "Addrs": ["/ip4/..."],
+}
+```
+  - `Keys` is a list of the CIDs being provided
+  - `Timestamp` is the current time
+  - `AdvisoryTTL` is the time by which the caller expects the server to keep the record available
+    - If this value is unknown, the caller may use a value of 0
+  - `ID` is the peer ID that was used to sign the record
+  - `Addrs` is a list of string-encoded multiaddrs
+
+A 403 response code should be returned if the signature check fails.
```

Comment:
`403 Forbidden` deals with client-side auth. Using is for signaling signature error on the server-side may be problematic if we ever have delegated routing endpoint behind some access control based on [Authorization](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization) header – client will display UI suggesting invalid Auth, hiding the fact auth was fine, but it was invalid signature.

Let's avoid this footgun and change this to generic [400 Bad Request](https://httpwg.org/specs/rfc9110.html#status.400):

```suggestion
A [400 Bad Request](https://httpwg.org/specs/rfc9110.html#status.400)  response code should be returned if the signature check fails.
```